rfcs/20220307-monitoring_across_the_stack.md

```diff
@@ -0,0 +1,353 @@
+
+The first one also seems complicated but I believe it's also easier. Remember that Prometheus is the source of truth for the configuration, so we wouldn't have to spread out the knowledge about where Prometheus lives in all our services. As long as our services are configured to serve Prometheus the appropriate data on consistent HTTP endpoints, then any Prometheus instance anywhere can consume them. That does mean that some configuration will be spread in the other direction: that is, rather than storing information about Prometheus in each service, each service's information will be stored in Prometheus. This seems like an appropriate compromise.
+
+I'm not sure where the best place to put the monitoring docker-compose is. I think I'd like to propose putting it in WordPress/openverse under a `monitoring` directory that would include the `docker-compose.yml` as well as documentation about how to configure monitors, our own best practices, etc. It doesn't make sense to me for us to just pick one of the three application repositories to put it in. The monitoring stack also fits in with the "meta"/"overview" nature of the WordPress/openverse repository.
+
+As for the cloud infrastructure side, I'll have to defer to Ronny, as I said above in the discussion about hosted vs self-hosted. It'd be nice if we could share configurations between local and development, and maybe there are some parts that we can, but overall due to the nature of how the production infrastructure is organized inside VPCs and such, I imagine there are probably some significant differences.
+
+Aside from service configuration, however, there is also the issue of how to store our Prometheus query, alerting, and other configuations locally. Likewise we need to address that for Grafana dashboards. For example, we don't want to have to manually copy the configurations for our anomaly monitors from production. Even if we develop these monitors in a live staging environment (more on that in the section below on the monitor development workflow) we'll still want to store those in code to have them committed to production, rather than manually copying them to prod. As long as the code they're saved to is accessible to the local monitoring stack, we should be able to seed the local stack with the same base monitors and dashboards present in production.
+
+### Local infrastructure recommendations
+
+One of the goals I'd like to achieve with the local infrastructure is to make it as easy as running `just monitor` and it spins up everything with tolerances for services that aren't currently running locally. For example, if someone is working primarily on developing frontend monitors, it would do no good for Prometheus to be producing errors about not be able to reach a local API, Airflow, image proxy, or Postgres instance. Same for someone working on the catalog, they should have to deal with answers about the frontend not being discoverable by Prometheus locally.
```

Comment:
Totally agree. Do you think it would be a reasonable requirement to have folks run `just monitor` from the Wordpress/openverse repository?

Reply:
If they're working on monitoring yes.

Reply:
Love this!